filters/OptimalNeighborhoodFilter.cpp

```diff
@@ -0,0 +1,166 @@
+/******************************************************************************
+ * Copyright (c) 2020, Bradley J Chambers (brad.chambers@gmail.com)
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following
+ * conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided
+ *       with the distribution.
+ *     * Neither the name of Hobu, Inc. or Flaxen Geo Consulting nor the
+ *       names of its contributors may be used to endorse or promote
+ *       products derived from this software without specific prior
+ *       written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include "OptimalNeighborhoodFilter.hpp"
+
+#include <pdal/KDIndex.hpp>
+
+#include <Eigen/Dense>
+
+#include <numeric>
+
+namespace pdal
+{
+using namespace Dimension;
+using namespace Eigen;
+
+static StaticPluginInfo const s_info{
+    "filters.optimalneighborhood", "OptimalNeighborhood Filter",
+    "http://pdal.io/stages/filters.optimalneighborhood.html"};
+
+CREATE_STATIC_STAGE(OptimalNeighborhood, s_info)
+
+std::string OptimalNeighborhood::getName() const
+{
+    return s_info.name;
+}
+
+OptimalNeighborhood::OptimalNeighborhood() : Filter() {}
+
+void OptimalNeighborhood::addArgs(ProgramArgs& args)
+{
+    args.add("min_k", "Minimum k-Nearest Neighbors", m_kMin, (point_count_t)10);
+    args.add("max_k", "Maximum k-Nearest Neighbors", m_kMax, (point_count_t)14);
+}
+
+void OptimalNeighborhood::addDimensions(PointLayoutPtr layout)
+{
+    m_kOpt = layout->registerOrAssignDim("OptimalKNN", Type::Unsigned64);
```

Comment:
These can be made standard.

Reply:
Will do.

Reply:
Fixed via https://github.com/PDAL/PDAL/pull/3180/commits/181546e9ce43e88caa401aaa70c4680bccf5f33c